c_api/include/taichi/taichi_cuda.h

```diff
@@ -0,0 +1,34 @@
+#pragma once
+#include <taichi/taichi_core.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif  // __cplusplus
+
+/*
+    We want to avoid including the entire <cuda.h> in C-API headers,
+    which exposes binary compatibility risks in the scenario where
+    the <cuda.h> used by C-API users is different from that used during taichi
+   compilation.
+
+    Thus we only copy neccessary declarations from <cuda.h>
+*/
+typedef void
+    *CUdeviceptr;  // Same implementation as "unsigned integer type matching
```

Comment:
I would expected the header to be included in `taichi.h` guarded by `#ifndef TI_NO_CUDA_INCLUDES`. Then the user can decide whether to include the header in `taichi.h` (I have to admit it was a problematic design but it's become a part of the convention 😭) or to declare the types themselves.

BTW, your comments will be flushed in a next run of `misc/generate_c_api.py`. Consider move them to `c_api/docs/taichi_cuda.h.md` and run `misc/generate_c_api_docs.py`. The CI will try to generate it and `diff` incoming commit outputs in the future.

Reply:
sounds good, thanks!